The Clear-Site-Data HTTP header should obey origin partition

cdumez opened the pull request:
#### 472954140c3524db1deebdf2a895d376548ebb52
<pre>
The Clear-Site-Data HTTP header should obey origin partition
<a href="https://bugs.webkit.org/show_bug.cgi?id=251094">https://bugs.webkit.org/show_bug.cgi?id=251094</a>

Reviewed by Youenn Fablet.

The Clear-Site-Data HTTP header should obey origin partition. If shouldn&apos;t be
possible for an iframe of origin A under top origin B to be able to clear site
data from top origin A (and vice-versa).

Our storages are partitioned and the request to clear site data should respect
that.

* LayoutTests/http/wpt/clear-site-data/partitioning-expected.txt: Added.
* LayoutTests/http/wpt/clear-site-data/partitioning.html: Added.
* LayoutTests/http/wpt/clear-site-data/resources/partitioning-popup.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/test_utils.sub.js:
* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::removeResourcesWithOrigin):
* Source/WebCore/loader/cache/MemoryCache.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::deleteCookies):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::deleteCookiesMatching):
(WebCore::NetworkStorageSession::deleteCookies):
(WebCore::NetworkStorageSession::deleteCookiesForHostnames):
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.cpp:
(WebCore::ServiceWorkerRegistrationKey::matchesOriginWithCachePartition const):
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::clear):
(WebCore::SWServer::clearInternal):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::deleteWebsiteDataForOrigin):
(WebKit::NetworkConnectionToWebProcess::deleteWebsiteDataForOrigins): Deleted.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataForOrigin):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::processClearSiteDataHeader):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::deleteData):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::deleteWebsiteDataForOrigin):
(WebKit::NetworkProcessConnection::deleteWebsiteDataForOrigins): Deleted.
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::deleteWebsiteDataForOrigin):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/259466@main">https://commits.webkit.org/259466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4f097863ef90906e77417cb96c3a7b616cd0a25

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104950 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14030 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114216 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174409 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108855 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4955 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97274 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113241 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110710 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11722 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39238 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26347 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7374 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27706 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7467 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4293 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13524 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47258 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6531 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9258 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -575,50 +577,59 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
+
+    RetainPtr<CFHTTPCookieStorageRef> cookieStorage = this->cookieStorage();
+    RetainPtr<NSHTTPCookieStorage> nsCookieStorage = adoptNS([[NSHTTPCookieStorage alloc] _initWithCFHTTPCookieStorage:cookieStorage.get()]);
+    auto aggregator = CallbackAggregator::create([completionHandler = WTFMove(completionHandler), nsCookieStorage] () mutable {
```
nsCookieStorage = WTFMove(nsCookieStorage)

cdumez replied:
Ok.

youennf commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -575,50 +577,59 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
+
+    RetainPtr<CFHTTPCookieStorageRef> cookieStorage = this->cookieStorage();
+    RetainPtr<NSHTTPCookieStorage> nsCookieStorage = adoptNS([[NSHTTPCookieStorage alloc] _initWithCFHTTPCookieStorage:cookieStorage.get()]);
+    auto aggregator = CallbackAggregator::create([completionHandler = WTFMove(completionHandler), nsCookieStorage] () mutable {
```
Should we use a main thread aggregator just in case?

youennf commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -575,50 +577,59 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
         return;
 
-    HashMap<String, Vector<RetainPtr<NSHTTPCookie>>> cookiesByDomain;
+    Vector<RetainPtr<NSHTTPCookie>> cookiesMatching;
```
cookiesMatching does not seem to be used.

cdumez replied:
Indeed, remains from a previous iteration. Will drop.

youennf commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -313,28 +313,42 @@ void SWServer::endSuspension()
 
 void SWServer::clear(const SecurityOriginData& securityOrigin, CompletionHandler<void()>&& completionHandler)
+{
+    clearInternal([securityOrigin](auto& key) {
```
&securityOrigin

cdumez replied:
I don't think this would be safe. clearInternal() can call the completionHandler asynchronously.

youennf commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -313,28 +313,42 @@ void SWServer::endSuspension()
+
+void SWServer::clear(const ClientOrigin& origin, CompletionHandler<void()>&& completionHandler)
+{
+    clearInternal([origin](auto& key) {
```
&origin

youennf commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -780,12 +781,17 @@ void NetworkResourceLoader::processClearSiteDataHeader(const WebCore::ResourceRe
 
     if (WebsiteDataStore::computeWebProcessAccessTypeForDataRemoval(typesToRemove, sessionID().isEphemeral()) != WebsiteDataStore::ProcessAccessType::None)
-        m_connection->deleteWebsiteDataForOrigins(typesToRemove, origins, [callbackAggregator] { });
+        m_connection->deleteWebsiteDataForOrigin(typesToRemove, clientOrigin, [callbackAggregator] { });
```
Why do we have to do this to that web process and not to all the other web processes?

cdumez replied:
Oh! this is not new to this patch but I think you're right. We should clear that data in all WebProcesses, not just the one that triggered this load. I'll fix.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -638,27 +638,39 @@ HashSet<WebCore::ClientOrigin> NetworkStorageManager::deleteDataOnDisk(OptionSet
-            return originSet.contains(origin.topOrigin) || originSet.contains(origin.clientOrigin);
-        });
+    m_queue->dispatch([this, protectedThis = Ref { *this }, types, originMatches = WTFMove(originMatches), completionHandler = WTFMove(completionHandler)]() mutable {
+        deleteDataOnDisk(types, -WallTime::infinity(), WTFMove(originMatches));
```
Is it safe to have originMatches be destroyed in a background thread?

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -638,27 +638,39 @@ HashSet<WebCore::ClientOrigin> NetworkStorageManager::deleteDataOnDisk(OptionSet
+    HashSet<WebCore::SecurityOriginData> originsToDelete;
+    originsToDelete.reserveInitialCapacity(origins.size());
+    for (auto origin : origins)
+        originsToDelete.add(origin.isolatedCopy());
```
OK, we are isolatedCopying, so we are ok, but it stills seems error prone.
I would be tempted to revert to the old behavior of explicitly isolateCopying.

cdumez replied:
It is safe as written but I agree it is fragile. I didn't find a good way to make the isolated copy more explicit without introducing more code duplication between the 2 functions.

It is not as easy as "reverting back to the old behavior", unless you mean keep the old function as-is, copy-paste it and tweak the new version to only on both topOrigin AND clientOrigin.

At the very least, I think I'll need to duplicate the `queue->dispatch()` call to make the isolated copying explicit.

cdumez replied:
Ok, I brought back the old code. It means a little bit a code duplication but it is less fragile.

youennf commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1450,6 +1451,16 @@ void WebProcess::deleteAllCookies(CompletionHandler<void()>&& completionHandler)
 
+void WebProcess::deleteWebsiteDataForOrigin(OptionSet<WebsiteDataType> websiteDataTypes, const ClientOrigin& origin, CompletionHandler<void()>&& completionHandler)
+{
+    if (websiteDataTypes.contains(WebsiteDataType::MemoryCache)) {
```
Should we ASSERT that websiteDataTypes.contains(WebsiteDataType::MemoryCache)?
Or are there cases where we send this message for no good reason?

cdumez replied:
I can add an assertion.

cdumez committed:
The Clear-Site-Data HTTP header should obey origin partition
https://bugs.webkit.org/show_bug.cgi?id=251094

Reviewed by Youenn Fablet.

The Clear-Site-Data HTTP header should obey origin partition. If shouldn't be
possible for an iframe of origin A under top origin B to be able to clear site
data from top origin A (and vice-versa).

Our storages are partitioned and the request to clear site data should respect
that.

* LayoutTests/http/wpt/clear-site-data/partitioning-expected.txt: Added.
* LayoutTests/http/wpt/clear-site-data/partitioning.html: Added.
* LayoutTests/http/wpt/clear-site-data/resources/partitioning-popup.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/clear-site-data/support/test_utils.sub.js:
* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::removeResourcesWithOrigin):
* Source/WebCore/loader/cache/MemoryCache.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::deleteCookies):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::deleteCookiesMatching):
(WebCore::NetworkStorageSession::deleteCookies):
(WebCore::NetworkStorageSession::deleteCookiesForHostnames):
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.cpp:
(WebCore::ServiceWorkerRegistrationKey::matchesOriginWithCachePartition const):
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::clear):
(WebCore::SWServer::clearInternal):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::deleteWebsiteDataForOrigin):
(WebKit::NetworkConnectionToWebProcess::deleteWebsiteDataForOrigins): Deleted.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataForOrigin):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::processClearSiteDataHeader):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::deleteData):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::deleteWebsiteDataForOrigin):
(WebKit::NetworkProcessConnection::deleteWebsiteDataForOrigins): Deleted.
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::deleteWebsiteDataForOrigin):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/259466@main
